query-pull/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+extern crate mentat_query;
+extern crate mentat_query_algebrizer;
+extern crate mentat_query_sql;
+extern crate mentat_sql;
+
+use std::collections::{
+    BTreeMap,
+    BTreeSet,
+};
+
+use std::iter::{
+    once,
+};
+
+use mentat_core::{
+    Cloned,
+    Entid,
+    HasSchema,
+    NamespacedKeyword,
+    Schema,
+    StructuredMap,
+    ValueRc,
+};
+
+use mentat_db::cache;
+
+use mentat_query::{
+    PullAttributeSpec,
+    PullConcreteAttribute,
+};
+
+pub mod errors;
+
+use errors::{
+    Result,
+};
+
+type PullResults = BTreeMap<Entid, ValueRc<StructuredMap>>;
+
+pub fn pull_attributes_for_entity<A>(schema: &Schema,
+                                     db: &rusqlite::Connection,
+                                     entity: Entid,
+                                     attributes: A) -> Result<StructuredMap>
+    where A: IntoIterator<Item=Entid> {
+    let attrs = attributes.into_iter()
+                          .map(|e| PullAttributeSpec::Attribute(PullConcreteAttribute::Entid(e)))
+                          .collect();
+    Puller::prepare(schema, attrs)?
+        .pull(schema, db, once(entity))
+        .map(|m| m.into_iter()
+                  .next()
+                  .map(|(k, vs)| {
+                      assert_eq!(k, entity);
+                      vs.cloned()
+                  })
+                  .unwrap_or_else(StructuredMap::default))
+}
+
+pub fn pull_attributes_for_entities<E, A>(schema: &Schema,
+                                          db: &rusqlite::Connection,
+                                          entities: E,
+                                          attributes: A) -> Result<PullResults>
+    where E: IntoIterator<Item=Entid>,
+          A: IntoIterator<Item=Entid> {
+    let attrs = attributes.into_iter()
+                          .map(|e| PullAttributeSpec::Attribute(PullConcreteAttribute::Entid(e)))
+                          .collect();
+    Puller::prepare(schema, attrs)?
+        .pull(schema, db, entities)
+}
+
+/// A `Puller` constructs on demand a map from a provided set of entity IDs to a set of structured maps.
+pub struct Puller {
+    // The domain of this map is the set of attributes to fetch.
+    // The range is the set of aliases to use in the output.
+    attributes: BTreeMap<Entid, ValueRc<NamespacedKeyword>>,
+    attribute_spec: cache::AttributeSpec,
+}
+
+impl Puller {
+    pub fn prepare_simple_attributes(schema: &Schema, attributes: Vec<Entid>) -> Result<Puller> {
+        Puller::prepare(schema,
+                        attributes.into_iter()
+                                  .map(|e| PullAttributeSpec::Attribute(PullConcreteAttribute::Entid(e)))
+                                  .collect())
+    }
+
+    pub fn prepare(schema: &Schema, attributes: Vec<PullAttributeSpec>) -> Result<Puller> {
+        // TODO: eventually this entry point will handle aliasing and that kind of
+        // thing. For now it's just a convenience.
+
+        let lookup_name = |i: &Entid| {
```

Comment:
Not in parsed representations, but in "pass it back in" scenarios.

E.g., ask Mentat for the attributes of some entity, then use those attributes when pulling another entity. We won't give you keywords, we'll give you entids, and it's cruel to make you turn those into keywords in order to invoke pull!

Also we don't support variable substitution inside pull expressions, so if they're dynamic I'd prefer people to inline/concatenate integers rather than arbitrary strings.

I'll switch this to err instead of composing a name.

Reply:
Ah, I see.  I still think that bailing for unnamed attributes is the right thing -- that's a violation of our mental model, and something that we should confront directly -- but I do agree that accepting entids for chaining queries is necessary.